pkg/workflow/tasks/discover.go

```diff
@@ -193,34 +195,29 @@ func (tr *stepGroupTaskRunner) Pending(ctx wfContext.Context, stepStatus map[str
 	return custom.CheckPending(ctx, tr.step, stepStatus)
 }
 
-func (tr *stepGroupTaskRunner) Skip(dependsOnPhase common.WorkflowStepPhase, stepStatus map[string]common.StepStatus) (common.StepStatus, bool) {
+// Run make workflow step group.
+func (tr *stepGroupTaskRunner) Run(ctx wfContext.Context, options *types.TaskRunOptions) (common.StepStatus, *types.Operation, error) {
 	status := common.StepStatus{
 		ID:   tr.id,
-		Name: tr.step.Name,
+		Name: tr.name,
 		Type: types.WorkflowStepTypeStepGroup,
 	}
-	if feature.DefaultMutableFeatureGate.Enabled(features.EnableSuspendOnFailure) {
-		return status, false
-	}
-	skip := custom.SkipTaskRunner(&custom.SkipOptions{
-		If:             tr.step.If,
-		DependsOnPhase: dependsOnPhase,
-	})
-	if skip {
-		status.Phase = common.WorkflowStepPhaseSkipped
-		status.Reason = custom.StatusReasonSkip
-		stepStatus[tr.step.Name] = common.StepStatus{
-			ID:    tr.id,
-			Phase: status.Phase,
+	for _, hook := range options.PreCheckHooks {
+		result, err := hook(tr.step)
+		if err != nil {
+			return common.StepStatus{}, nil, errors.WithMessage(err, "do preCheckHook")
+		}
+		if result.Skip {
+			status.Phase = common.WorkflowStepPhaseSkipped
+			status.Reason = custom.StatusReasonSkip
+			options.StepStatus[tr.step.Name] = status
+			break
+		} else if result.Timeout {
```

Comment:
```suggestion
		}
		if result.Timeout {
```

We don't need to use else as the former case will always break